### PR TITLE
feat: filter changes

### PR DESCRIPTION
**Pull request by anandtiwary**
## Description
Please include a summary of the change, motivation and context.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: projects/components/src/filtering/filter-button/filter-button.component.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export class FilterButtonComponent implements OnChanges {
   public attribute?: FilterAttribute;
 
   @Input()
-  public value?: unknown;
+  public value?: FilterValue;
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter-button/filter-button.component.ts:48]: Seems completely dependent on the filter implementation - what's the point of typing it?

File: projects/components/src/filtering/filter/filter-operators.ts [hunk 2]
```diff
@@ -9,6 +9,7 @@ export const enum FilterOperator {
   GreaterThanOrEqualTo = '>=',
   Like = '~',
   In = 'IN',
+  NotIn = 'NOT_IN',
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/filter-operators.ts:12]: Pretty sure this isn't supported, if it's being added please put in a separate PR

File: projects/components/src/filtering/filter/filter-operators.ts [hunk 3]
```diff
@@ -44,6 +46,8 @@ export const toUrlFilterOperator = (operator: FilterOperator): UrlFilterOperator
       return UrlFilterOperator.In;
     case FilterOperator.ContainsKey:
       return UrlFilterOperator.ContainsKey;
+    case FilterOperator.NotIn:
+      return UrlFilterOperator.ContainsKey;
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/filter-operators.ts:50]: bad map

File: projects/components/src/filtering/filter/filter-operators.ts [hunk 4]
```diff
@@ -78,6 +84,7 @@ export const incompatibleOperators = (operator: FilterOperator): FilterOperator[
   switch (operator) {
     case FilterOperator.In:
     case FilterOperator.Equals:
+    case FilterOperator.NotIn:
       return [
         FilterOperator.In,
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/filter-operators.ts:89]: this logic seems off and is not symmetric. `Not in` would return `in` is incompatible, but `in` would not return `not in` as incompatible.

File: projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts [hunk 5]
```diff
@@ -7,6 +7,7 @@ export interface AttributeMetadata {
   units: string;
   type: AttributeMetadataType;
   scope: string;
+  category?: string;
```
Comment (aaron-steinfeld) [inline projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts:10]: not generally applicable nor part of GQL model, please remove

File: projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts [hunk 6]
```diff
@@ -1,13 +1,36 @@
...
-import { FilterOperator, TableFilter } from '@hypertrace/components';
+import { FilterOperator } from '@hypertrace/components';
 import { GraphQlArgumentValue } from '@hypertrace/graphql-client';
 import { GraphQlFieldFilter } from '../../graphql/model/schema/filter/field/graphql-field-filter';
 import { GraphQlFilter, GraphQlOperatorType } from '../../graphql/model/schema/filter/graphql-filter';
 
 @Injectable({ providedIn: 'root' })
 export class GraphQlFilterBuilderService {
-  public buildGraphQlFilters(filters: TableFilter[]): GraphQlFilter[] {
+  public buildFiltersFromGraphQlFieldFilters(filters: GraphQlFieldFilter[]): FieldFilter[] {
```
Comment (aaron-steinfeld) [inline projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts:12]: Why do we want to convert back from GQL filters to field filters? If we're doing refactoring, we should change any components using GQL filters to use the UI construct instead of supporting going bidrectionally.

File: projects/components/src/filtering/filter-button/filter-button.component.ts (hunk 1 above)
Comment (anandtiwary) [reply projects/components/src/filtering/filter-button/filter-button.component.ts:48]: This is required for Filter to GqlFilter transformation

File: projects/components/src/filtering/filter/filter-operators.ts (hunk 2 above)
Comment (anandtiwary) [reply projects/components/src/filtering/filter/filter-operators.ts:12]: I saw one operator in Gql. So thought it is supported. 

yes, this is just a draft PR. I will create smaller ones when it is ready

File: projects/components/src/filtering/filter/filter-operators.ts (hunk 4 above)
Comment (anandtiwary) [reply projects/components/src/filtering/filter/filter-operators.ts:89]: Yeah, I am not sure what this is code is doing. Still playing around with it.

File: projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts (hunk 5 above)
Comment (anandtiwary) [reply projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts:10]: The category would be added to attributes eventually, right?

File: projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts (hunk 6 above)
Comment (anandtiwary) [reply projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts:12]: Yeah, so that would be a big change. If we want to do it in iterations then temporarily this method will be used in the basic filter bar for now. Slowly, we will migrate the filters to the UI construct. 

We would also need to revisit the relationship between `IncompleteFilter`, `FieldFilter`, and `Filter`?

File: projects/components/src/filtering/filter/filter-operators.ts (hunk 3 above)
Comment (anandtiwary) [reply projects/components/src/filtering/filter/filter-operators.ts:50]: yeah. Do you know we need three operator types? GqlOperators, FilterOperator and URLFilterOperator. Imo We can combine FilterOperator and URLFilterOperator and only use one.

File: projects/components/src/filtering/filter-button/filter-button.component.ts (hunk 1 above)
Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter-button/filter-button.component.ts:48]: This is the kind of thing I was saying the other day we shouldn't try to make generic, because there is no one GQL filter representation.

File: projects/components/src/filtering/filter/filter-operators.ts (hunk 3 above)
Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter/filter-operators.ts:50]: There's three different string representations of a user - the server's value (e.g. `EQUALS`), the value we display in the UI (`=`) and the value we use in a URL string where we can't use all specials chars (`_eq_`). We probably could try to consolidate such as by url encoding or similar, but seemed easy enough to maintain initially.

File: projects/components/src/filtering/filter/filter-operators.ts (hunk 4 above)
Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter/filter-operators.ts:89]: I'm not 100%, but I believe this is used when adding filters via table controls, so if you do something like add an equals filter it replaces any current one.

File: projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts (hunk 5 above)
Comment (aaron-steinfeld) [reply projects/observability/src/shared/graphql/model/metadata/attribute-metadata.ts:10]: I don't believe so - I thought the conclusion there was that concept was unique to a particular UI page and could be maintained there. It's not a universal concept across all attributes, nor is it necessarily universal that the same attribute belongs to the same category in all contexts - its use case specific.

File: projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts (hunk 6 above)
Comment (anandtiwary) [reply projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts:13]: Things I am re-using:
1. Filter Objects and various filter builders 
2. Url string to Filter parsing logic
3. Basic filter bar value auto-completion

Things I am changing:
1. Moving UI side code to depend on `Filter` objects instead of GraphQlFilters. We will use a method that would have the conversion logic. This migration will have in phases and we can track it with stories.
2. Fix the dependency of Filter, FieldFilter, and IncompleteFilter
2. Using new design for basic view


@jake-bassett Migrating to UI construct (point 1) has a much bigger blast radius. We should do it in stages. Let's wait for other higher-priority tasks to get completed first.

Comment (jake-bassett) [reply projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts:13]: Okay, got it. Yeah, the items you are re-using are in desperate need of rework but I know that's not at all trivial. I'm good with doing this in phases. As much as I know that code is hard to work with, and making things a bit of a challenge, I don't think we can wait on these filter changes much longer.

I vote for not making this PR any larger that what it already is and revisiting replacing the rest of the filtering at a later date. Would appreciate if @aaron-steinfeld can chime in on any more quick wins we can do here in the meantime though.

Comment (anandtiwary) [reply projects/observability/src/shared/services/filter-builder/graphql-filter-builder.service.ts:13]: okay. approve this PR  :)

**Review by jake-bassett — APPROVED** [review]
